### DISCUSSION TITLE
Drawing using LVGL

### DISCUSSION THREAD

**Discussion by polarkernel** [rzeldent/esp32-smartdisplay#218] [Q&A] [answered]
I would like to thank René for his great work, and I really enjoy working with this library!

However, in comparison to the external libraries mentioned ([TFT_eSPI](https://github.com/Bodmer/TFT_eSPI) or [LovyanGFX](https://github.com/lovyan03/LovyanGFX)), I miss the ability to do drawing. For instance TFT_eSPI provides functions like

```
drawPixel(), drawLine(), drawRect(), drawCircle(), drawTriangle(), drawRoundRect()
fillScreen(), fillRect(), fillCircle(), fillTriangle()
```

Using the additional library [TFTShape](https://github.com/androdlang/TFTShape) of _androdlang_, even polylines, polygons, grids, splines, stars and n-gons can be drawn using TFT_eSPI. Since LVGL version 9.x is now integrated into esp32-smartdisplay (another big step forward, thanks to René), [ThorVG](https://github.com/thorvg/thorvg), which is now part of LVGL, also contains vector-oriented drawing functions. However, I did not manage to squeeze this library into the existing memory.

My question is now: Has anyone been able to successfully use ThorVG on one of the Sunton boards? Or does anyone have any knowledge of where drawing functions exist or could be created that are based on LVGL? My goal is to draw vector data from OpenStreetMap for a GPS application. I would need to draw polylines of different width and color and filled polygons (preferably also concave ones) of different colors.

I know my question is kind of off-topic here, I hope you forgive me.

**Comment by rzeldent** [comment]
Hi Polarkernel,

I believe SVG is supported in LVGL. SVG has all these vector definitions and this might be a solution!

**Comment by polarkernel** [comment]
Hi René

I always thought that SVG was included in ThorVG. After your comment, I found out that there is a separate library _libs/svg_ in lvgl version 9.2.2 on [GitHub](https://github.com/lvgl/lvgl). But the lvgl code automatically loaded by your library doesn't load this library and also doesn't load the corresponding template file _lv_conf_template.h_. The loaded template file does say to be _Configuration file for v9.2.2_, but the definition `LV_USE_SVG` doesn't exist there. Any idea where this difference comes from?

Thanks for your hint, I will try to install this library manually, eventually, it could easily solve my problem.



> However, I did not manage to squeeze this library into the existing memory.

I tried to implement a [given example](https://github.com/lvgl/lvgl/blob/master/examples/widgets/canvas/lv_example_canvas_8.c), but the linker wasn't able to fit it into the memory (4 MB) of my board (esp32-2432W328C):

<img src="https://github.com/user-attachments/assets/746185d4-3610-4c93-8f27-4473f3b89d49" width="800">

That is what I meant by this sentence in my initial question.

**Comment by polarkernel** [comment]
It is strange, I am not able to load the SVG library. The version 9.2.2. of LVLG that _platformio_ loads when I use _smartdisplay_ is different from the version provided on [GitHub](https://github.com/lvgl/lvgl). The _lvgl/src/libs/svg_ library does not exist. When I replace the LVGL code with the GitHub version (when VS Code is running) and then try to build my program, this library is automatically replaced with the old version (by _platformio_, I assume):
```
Library Manager: Installing lvgl/lvgl @ ^9.2.2
Unpacking  [####################################]  100%
Library Manager: lvgl@9.2.2 has been installed!
```
I have no idea what the origin of this version is.

I give up for now. It is a pity that LVGL does not provide primitive functions like `drawPixel(x,y,color)`, so that a user could develop his own custom widgets. I could easily write my own graphics library if I had an efficient solution for such primitives. For example, the graphics functions of TFT_eSPI are based on similar functions.

**Comment by rzeldent** [comment]
Well, there are also functions like:
- https://docs.lvgl.io/master/details/widgets/line.html
- https://docs.lvgl.io/master/details/widgets/arc.html
- https://docs.lvgl.io/master/details/widgets/bar.html

So this can also be done using LVGL!

**Comment by polarkernel** [comment]
> So this can also be done using LVGL!

I already knew that, thanks, and I already had a working version using `lv_line`. But `lv_line`, for example, draws lines of arbitrary width, applying antialiasing and drawing rounded ends. Beautiful! But it can't be efficient. For my GPS application it might work, because drawings change slowly there. 

It seems that vector drawing is implemented in version 9.2.2 of LVGL. However, I cannot access it. It irritates me that the version of LVGL loaded by _platformio_ and the version on GitHub are different, although they show the same version number. However, since my knowledge of the hardware is also very limited, I do not want to evaluate it more deeply. Maybe it will be fixed in future versions.

**Comment by rzeldent** [comment]
@polarkernel You can use the github version of LVGL by changing this in the library:

[library.json](https://github.com/rzeldent/esp32-smartdisplay/blob/main/library.json). Here change the dependencies to the github repo!

BTW: It is already LVGL 9.2.2 Maybe something needs to be included in the lvgl configuration?

**Comment by polarkernel** [comment, marked as answer]
> BTW: It is already LVGL 9.2.2 Maybe something needs to be included in the lvgl configuration?

I think I found the reason for the observed difference. I get LVGL version 9.2.2, when I build with one of the following methods, which do the same for LVGL:
```
; either with
lib_deps = 
	rzeldent/esp32_smartdisplay @2.1.0

; or with
lib_deps = 
	lvgl @^9.2.2
```
But when I build with 

```
lib_deps = 
	https://github.com/lvgl/lvgl.git
```
then the library manager reports:
```
        Library Manager: lvgl@9.3.0-dev+sha.df18a5f has been installed!
        Updating metadata for the vscode IDE...
```
This means, that the master branch of LVGL on GitHub is the development version 9.3.0 and not 9.2.2, which is the release version. SVG is not yet implemented in version 9.2.2, but in the new version, which has not yet been released. This knowledge scares the spirits away and the fog is lifted!
